lib/class-wp-rest-block-renderer-controller.php

```diff
@@ -0,0 +1,128 @@
+<?php
+/**
+ * Block Renderer REST API: WP_REST_Block_Renderer_Controller class
+ *
+ * @package gutenberg
+ * @since ?
+ */
+
+/**
+ * Controller which provides REST endpoint for rendering a block.
+ *
+ * @since ?
+ *
+ * @see WP_REST_Controller
+ */
+class WP_REST_Block_Renderer_Controller extends WP_REST_Controller {
+
+	/**
+	 * Constructs the controller.
+	 *
+	 * @access public
+	 */
+	public function __construct() {
+		// @codingStandardsIgnoreLine - PHPCS mistakes $this->namespace for the namespace keyword.
+		$this->namespace = 'gutenberg/v1';
+		$this->rest_base = 'block-renderer';
+	}
+
+	/**
+	 * Registers the necessary REST API routes, one for each dynamic block.
+	 *
+	 * @access public
+	 */
+	public function register_routes() {
+		$block_types = WP_Block_Type_Registry::get_instance()->get_all_registered();
+		foreach ( $block_types as $block_type ) {
+			if ( ! $block_type->is_dynamic() ) {
+				continue;
+			}
+
+			// @codingStandardsIgnoreLine - PHPCS mistakes $this->namespace for the namespace keyword.
+			register_rest_route( $this->namespace, '/' . $this->rest_base . '/(?P<name>' . $block_type->name . ')', array(
+				'args'   => array(
+					'name' => array(
+						'description' => __( 'Unique registered name for the block.', 'gutenberg' ),
+						'type'        => 'string',
+					),
+				),
+				array(
+					'methods'             => WP_REST_Server::READABLE,
+					'callback'            => array( $this, 'get_item' ),
+					'permission_callback' => array( $this, 'get_item_permissions_check' ),
+					'args'                => array(
+						'context'    => $this->get_context_param( array( 'default' => 'view' ) ),
+						'attributes' => array(
+							/* translators: %s is the name of the block */
+							'description'          => sprintf( __( 'Attributes for %s block', 'gutenberg' ), $block_type->name ),
+							'type'                 => 'object',
+							'additionalProperties' => false,
+							'properties'           => $block_type->attributes,
+						),
+					),
+				),
+				'schema' => array( $this, 'get_public_item_schema' ),
+			) );
+		}
+	}
+
+	/**
+	 * Checks if a given request has access to read blocks.
+	 *
+	 * @since ?
+	 * @access public
+	 *
+	 * @param WP_REST_Request $request Request.
+	 * @return true|WP_Error True if the request has read access, WP_Error object otherwise.
+	 */
+	public function get_item_permissions_check( $request ) {
+		if ( ! current_user_can( 'edit_posts' ) ) {
```

Comment:
Something that comes to mind is that some blocks will want to render differently depending on their post context. In that case, there should be an `edit_post` check for a given post ID.

At least this is the case for shortcodes, for example a `[gallery]` will render the images attached to the current post if no `ids` attribute is present. A similar use case seems similar for dynamic blocks. This could be particularly powerful for re-usable blocks. For instance a nav menu (#5036) will render a menu item differently when that item is the one currently being viewed. When a dynamic block is used as a reusable block this would allow for its post context to be variable.

The author bio block in #3250 is another example for where for there to be a post context to render.

Should the post dependency be formalized as a new [attribute](https://github.com/WordPress/gutenberg/blob/master/docs/attributes.md) source type to make the explicit and declarative that the block depends on a post context to be present?

In the case of the author bio block, this could be formalized as follows instead of relying on `getEditedPostAttribute` directly:

```php
attributes: {
	author: {
		type: 'int',
		source: 'post-context',
		field: 'author'
	},
},
```

This would allow for the attribute's source to be matched in a server-side context as well.

Reply:
@westonruter To see if I understand the thoughts correctly:
Are you bringing out the issue that depending on the block the `edit_posts` permission might not be enough since the block might involve other posts? For example in case of nav menus, we should check if the user has the sufficient permissions for the nav menu items? And in case of shortcodes (e.g. gallery) if the user can edit the specific post (check by existing `ID`/ `postId` in the request)?

On the new `source` -- not sure I understand completely. Are you thinking that the `post-context` could be used in case we need to get the object data loaded ( e.g. in this case author user object )? Or would it just be to note that the block depends on the author?

Reply:
I commented on the `edit_posts` cap check because it was related but not the primary concern. The first concern is that a dynamic block may depend on a post for context in order to display properly. For example, the aforementioned author bio block requires a `$post->post_author` global context to display. In the same way, the `gallery` widget depends on a global `$post` to find attachments to display if there are no `ids` present. Blocks could reply on postmeta to display, but if there is no `$post` context then it wouldn't know from where to get the postmeta.

I see this being analogous to the [`parse-embed` Ajax action](https://github.com/WordPress/wordpress-develop/blob/96695f28644043fa47cb6843e3a29a7959a00bd4/src/wp-admin/includes/ajax-actions.php#L3258-L3273). Notice how it allows for a post to be supplied in the request for global context.

So then the topic of the new attribute `source` type is to formalize the relationship between a block to having a post as context. In a server-side rendering context, this would normally be implemented by setting the `$post` global. In a client-side context, it could be realized as [shown in the author bio block](https://github.com/WordPress/gutenberg/pull/3250/files#diff-cbc857512e8990d9e6b39655c87fd7c9R99):

```js
postAuthorId: getEditedPostAttribute( state, 'author' ),
```

The impact of the `$post` context could also be directly visible in the nav menu block (#5036) since nav menu items have classes that vary based on the `$post` context, as can be seen in core's `_wp_menu_item_classes_by_context` function.

Reply:
@westonruter Added setting the global `$post` in case the param is set, let me know if you have any thoughts on what else might be missing from the SSR component side.